cpp/custrings/strings/attrs.cu

```diff
@@ -437,3 +438,43 @@ unsigned int NVStrings::is_empty( bool* results, bool todevice )
     return (unsigned int)matches;
 }
 
+//
+// s = ["a","xyz", "éee"]
+// s.code_points(results)
+// results is [   97   120   121   122 50089   101   101]
+//
+unsigned int NVStrings::code_points( unsigned int* d_results )
+{
+    auto count = size();
+    auto execpol = rmm::exec_policy(0);
+    custring_view_array d_strings = pImpl->getStringsPtr();
+
+    // offsets point to each individual integer range
+    rmm::device_vector<size_t> offsets(count);
+    size_t* d_offsets = offsets.data().get();
+    thrust::transform_inclusive_scan(execpol->on(0),
+        thrust::make_counting_iterator<size_t>(0),
+        thrust::make_counting_iterator<size_t>(count),
+        d_offsets,
+        [d_strings] __device__(size_t idx){
+            custring_view* d_str = d_strings[idx];
+            size_t length = 0;
+            if( d_str )
+                length = d_str->chars_count();
+            return length;
+        },
+        thrust::plus<unsigned int>());
+
+    // now set the ranges from each strings' character values
+    thrust::for_each_n(execpol->on(0), thrust::make_counting_iterator<unsigned int>(0), count,
+        [d_strings, d_offsets, d_results] __device__(unsigned int idx){
+            custring_view* d_str = d_strings[idx];
+            if( !d_str )
+                return;
+            auto result = d_results + (idx ? d_offsets[idx-1] : 0);
+            for( auto itr = d_str->begin(); itr != d_str->end(); ++itr )
```

Comment:
Replace this loop with a sequential `copy()`.

```suggestion
thrust::copy(thrust::seq(), d_str->begin(), d_str->end(), result); 
```

Reply:
I wonder if you may be able to parallelize over the entire output size rather than doing a sequential copy for each string.  Is `d_strings` a contiguous array of bytes? Would it be possible to do the `for_each` over the total size of the results array instead of over the number of strings? If so, you could get more parallelism and avoid the variable-size-per-thread sequential loop that's currently there.

Reply:
The `d_strings` is not contiguous array of bytes.

Reply:
`thrust::copy()` does not recognize the `custring_view::iterator` because it requires a `difference_type`. This is a special iterator for quickly navigating the variable-length characters and so has some limitations like this.

Reply:
Adding a `difference_type` is pretty trivial.

Reply:
Not for variable-width characters. I don't think it is worth doing.

Reply:
It's definitely worth having an iterator based on standards. A key reason for the existence of iterators is composability with standard algorithms and containers.

Reply:
I can certainly try to build one in the newer version of this class being created here:
https://github.com/rapidsai/cudf/blob/425c81f177c832dad3fc09ed10029714e80c36b7/cpp/include/cudf/strings/string_view.cuh#L86